File: plugins/humanoverflow/scripts/inbox.py
#!/usr/bin/env python3
"""Read/mark helpers for the humanoverflow daemon inbox.

The daemon (scripts/hof-mcp.py daemon) appends incoming questions to
~/.cache/humanoverflow/inbox/<project-key>.jsonl. This script gives skills
a stable interface to query unprocessed entries and mark them done, without
each skill having to know the on-disk layout.

Subcommands:
  inbox.py list [--limit N]        unprocessed entries as JSON lines
  inbox.py mark <questionId>...    append IDs to the processed sidecar
  inbox.py status                  counts + paths (for status skills)

The processed sidecar is append-only (one questionId per line), so the
daemon's append-only inbox and the skill's append-only sidecar never race.
"""
from __future__ import annotations

import argparse
import hashlib
import json
import os
import subprocess
import sys
from pathlib import Path

CACHE_DIR = Path.home() / ".cache" / "humanoverflow"
INBOX_DIR = CACHE_DIR / "inbox"


def git_toplevel() -> Path:
    try:
        out = subprocess.run(
            ["git", "rev-parse", "--show-toplevel"],
            capture_output=True, text=True, check=True, timeout=5,
        )
        return Path(out.stdout.strip())
    except (subprocess.CalledProcessError, FileNotFoundError, subprocess.TimeoutExpired):
        return Path.cwd()


def project_key(root: Path | None = None) -> str:
    root = root or git_toplevel()
    return hashlib.sha256(str(root.resolve()).encode()).hexdigest()[:16]


def paths(key: str) -> tuple[Path, Path]:
    return INBOX_DIR / f"{key}.jsonl", INBOX_DIR / f"{key}.processed"


def load_processed_ids(processed: Path) -> set[str]:
    if not processed.is_file():
        return set()
    return {line.strip() for line in processed.read_text().splitlines() if line.strip()}


def iter_inbox(inbox: Path):
    if not inbox.is_file():
        return
    with open(inbox) as fh:
        for line in fh:
            line = line.strip()
            if not line:
                continue
            try:
                yield json.loads(line)
            except json.JSONDecodeError:
                continue


def cmd_list(args: argparse.Namespace) -> int:
    key = project_key()
    inbox, processed = paths(key)
    done = load_processed_ids(processed)
    count = 0
    for q in iter_inbox(inbox):
        qid = q.get("id")
        if qid in done:
            continue
        sys.stdout.write(json.dumps(q, ensure_ascii=False) + "\n")
        count += 1
        if args.limit and count >= args.limit:
            break
    return 0


def cmd_mark(args: argparse.Namespace) -> int:
    key = project_key()
    _inbox, processed = paths(key)
    processed.parent.mkdir(parents=True, exist_ok=True)
    with open(processed, "a") as fh:
        for qid in args.ids:
            fh.write(qid + "\n")
    return 0


def cmd_status(_args: argparse.Namespace) -> int:
    key = project_key()
    inbox, processed = paths(key)
    total = sum(1 for _ in iter_inbox(inbox))
    done = load_processed_ids(processed)
    unprocessed = sum(1 for q in iter_inbox(inbox) if q.get("id") not in done)
    print(json.dumps({
        "project_key": key,
        "inbox": str(inbox),
        "processed_sidecar": str(processed),
        "total": total,
        "processed": len(done),
        "unprocessed": unprocessed,
    }, indent=2))
    return 0


def main() -> int:
    parser = argparse.ArgumentParser(prog="inbox.py", description=__doc__)
    sub = parser.add_subparsers(dest="cmd", required=True)

    pl = sub.add_parser("list", help="emit unprocessed inbox entries as JSON lines")
    pl.add_argument("--limit", type=int, default=0, help="max entries to emit (0 = no limit)")

    pm = sub.add_parser("mark", help="append questionId(s) to the processed sidecar")
    pm.add_argument("ids", nargs="+")

    sub.add_parser("status", help="counts + paths")

    args = parser.parse_args()
    if args.cmd == "list":
        return cmd_list(args)
    if args.cmd == "mark":
        return cmd_mark(args)
    if args.cmd == "status":
        return cmd_status(args)
    return 2


if __name__ == "__main__":
    sys.exit(main())

File: plugins/agentnetwork/scripts/inbox.js
#!/usr/bin/env node
// Read/mark helpers for the agentnetwork daemon inbox.
//
// The daemon (scripts/an-mcp.js daemon) appends incoming questions to
// ~/.cache/agentnetwork/inbox/<project-key>.jsonl. This script gives skills
// a stable interface to query unprocessed entries and mark them done, without
// each skill having to know the on-disk layout.
//
// Subcommands:
//   inbox.js list [--limit N]        unprocessed entries as JSON lines
//   inbox.js mark <questionId>...    append IDs to the processed sidecar
//   inbox.js status                  counts + paths (for status skills)
//
// The processed sidecar is append-only (one questionId per line), so the
// daemon's append-only inbox and the skill's append-only sidecar never race.
//
// Cross-platform: Node 20+ stdlib only. Works on Linux, macOS, Windows.

'use strict';

const crypto = require('node:crypto');
const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');
const { spawnSync } = require('node:child_process');

const CACHE_DIR = path.join(os.homedir(), '.cache', 'agentnetwork');
const INBOX_DIR = path.join(CACHE_DIR, 'inbox');

function gitToplevel() {
  const r = spawnSync('git', ['rev-parse', '--show-toplevel'], {
    encoding: 'utf8',
    timeout: 5000,
  });
  if (r.status === 0 && r.stdout) return r.stdout.trim();
  return process.cwd();
}

function projectKey(root) {
  const abs = path.resolve(root || gitToplevel());
  return crypto.createHash('sha256').update(abs).digest('hex').slice(0, 16);
}

function paths(key) {
  return {
    inbox: path.join(INBOX_DIR, `${key}.jsonl`),
    processed: path.join(INBOX_DIR, `${key}.processed`),
  };
}

function loadProcessedIds(processed) {
  if (!fs.existsSync(processed) || !fs.statSync(processed).isFile()) {
    return new Set();
  }
  const text = fs.readFileSync(processed, 'utf8');
  const out = new Set();
  for (const line of text.split('\n')) {
    const s = line.trim();
    if (s) out.add(s);
  }
  return out;
}

function* iterInbox(inbox) {
  if (!fs.existsSync(inbox) || !fs.statSync(inbox).isFile()) return;
  const text = fs.readFileSync(inbox, 'utf8');
  for (const line of text.split('\n')) {
    const s = line.trim();
    if (!s) continue;
    try {
      yield JSON.parse(s);
    } catch (_) {
      // skip malformed line, mirrors python json.JSONDecodeError path
    }
  }
}

function cmdList(args) {
  const key = projectKey();
  const { inbox, processed } = paths(key);
  const done = loadProcessedIds(processed);
  let count = 0;
  for (const q of iterInbox(inbox)) {
    const qid = q && q.id;
    if (qid && done.has(qid)) continue;
    process.stdout.write(JSON.stringify(q) + '\n');
    count += 1;
    if (args.limit && count >= args.limit) break;
  }
  return 0;
}

function cmdMark(args) {
  const key = projectKey();
  const { processed } = paths(key);
  fs.mkdirSync(path.dirname(processed), { recursive: true });
  const fd = fs.openSync(processed, 'a');
  try {
    for (const qid of args.ids) fs.writeSync(fd, qid + '\n');
  } finally {
    fs.closeSync(fd);
  }
  return 0;
}

function cmdStatus() {
  const key = projectKey();
  const { inbox, processed } = paths(key);
  let total = 0;
  for (const _ of iterInbox(inbox)) total += 1;
  const done = loadProcessedIds(processed);
  let unprocessed = 0;
  for (const q of iterInbox(inbox)) {
    if (!q || !done.has(q.id)) unprocessed += 1;
  }
  process.stdout.write(
    JSON.stringify(
      {
        project_key: key,
        inbox,
        processed_sidecar: processed,
        total,
        processed: done.size,
        unprocessed,
      },
      null,
      2,
    ) + '\n',
  );
  return 0;
}

function parseArgs(argv) {
  const [cmd, ...rest] = argv;
  if (!cmd) return { cmd: null };
  if (cmd === 'list') {
    let limit = 0;
    for (let i = 0; i < rest.length; i++) {
      if (rest[i] === '--limit') {
        limit = parseInt(rest[++i] || '0', 10) || 0;
      } else if (rest[i].startsWith('--limit=')) {
        limit = parseInt(rest[i].slice('--limit='.length), 10) || 0;
      }
    }
    return { cmd, limit };
  }
  if (cmd === 'mark') {
    if (rest.length === 0) {
      process.stderr.write('mark: at least one questionId required\n');
      return { cmd: null };
    }
    return { cmd, ids: rest };
  }
  if (cmd === 'status') return { cmd };
  process.stderr.write(`unknown subcommand: ${cmd}\n`);
  return { cmd: null };
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  if (!args.cmd) {
    process.stderr.write(
      'usage: inbox.js list [--limit N] | mark <id>... | status\n',
    );
    return 2;
  }
  if (args.cmd === 'list') return cmdList(args);
  if (args.cmd === 'mark') return cmdMark(args);
  if (args.cmd === 'status') return cmdStatus();
  return 2;
}

process.exit(main());
